### PR TITLE
Add apex tests for signing

### PR DESCRIPTION
**Pull request by PatoBeltran**
Fix https://github.com/NuGet/Home/issues/6275
Added some end to end scenarios for package signing in Apex

### CODE REVIEW

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (emgarten) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: `.Result`  should be the same

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (emgarten) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: This is a long wait for something that is tested for 3 templates. I would either make this for 1 template, or create just 1 expired package and re-use it for all tests.

**Review by emgarten — CHANGES_REQUESTED** [review]

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (jainaashish) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: what happens when it fails to create a sign package?

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (jainaashish) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: feels redundant, you already added a new test for signed package and there's an existing test for unsigned package. Are we getting any extra coverage with this test? Its good to have as many tests as possible but avoid duplicates.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (jainaashish) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: same as above

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (jainaashish) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: This again feels over-kill. IMO we should try and cover all different scenarios/ edge cases via unit tests or func tests. But have E2E tests for the mainline scenarios just to make sure it works all the way end to end. We shouldn't need to cover everything through E2E tests. It will make it really slow to run all E2E tests over time.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (jainaashish) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: why are we testing this when we anyways install this package with expired certificate?

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (jainaashish) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: how do we know the certificate got expired?

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (jainaashish) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: we should also confirm that installation failed because signed package was tampered.

**Review by jainaashish — CHANGES_REQUESTED** [review]
Please check if we really need to add all these E2E tests and that too each scenario for all the different project templates?

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (dtivel) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: I think it would be easier and more robust to create a package with a certificate that will expire immediately after signing and then add that signed package as a static test resource.  If there's a problem with trust you could extract the certificate chain and temporarily add the root as a trusted root authority.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (PatoBeltran) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: You are right. The test cases I will test will be:
- Install signed package
- Uninstall signed package
- Update from unsiged to signed
- Downgrade from signed to unsigned
- Install tampered signed
- Install untrusted (e.g. expired cert) signed

**Comment by PatoBeltran** [comment]
@emgarten @jainaashish I updated the PR removing some duplicate tests and reusing more code, please review.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (jainaashish) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: As discussed, let have two project templates, c# class library and netcore console app.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (jainaashish) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: Why are we allowing to update from a signed package to unsigned package? I don't see any logical reason for this.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (jainaashish) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: There should be some warning shown to the consumer about expired certificate.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (dtivel) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs]: It's not ideal, but that's the current state of the product and it's fine to capture current non-ideal product state in a test.  The current behavior even with unsigned packages is "first package source wins."  I believe signed packages should have special treatment in that they should not be silently updated to an unsigned version or a different signer, but also depending on the current effective client policy.  We're not there yet.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/Fixtures/IVisualStudioHostFixtureFactory.cs
Comment (dtivel) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/Fixtures/IVisualStudioHostFixtureFactory.cs]: license header?

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/Fixtures/SignedPackagesTestsApexFixture.cs
Comment (dtivel) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/Fixtures/SignedPackagesTestsApexFixture.cs:4]: license header?

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/Fixtures/VisualStudioHostFixtureFactory.cs
Comment (dtivel) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/Fixtures/VisualStudioHostFixtureFactory.cs]: license header?

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/Fixtures/SignedPackagesTestsApexFixture.cs
Comment (dtivel) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/Fixtures/SignedPackagesTestsApexFixture.cs]: This assumes all steps between certificate generation and signing took less than 5 seconds.  I think you should assert after this sleep that the signing certificate is now expired.

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/Fixtures/SignedPackagesTestsApexFixture.cs
Comment (dtivel) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/Fixtures/SignedPackagesTestsApexFixture.cs:4]: Check other files too.  I didn't comment on them all.

**Review by dtivel — APPROVED** [review]

File: test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestPackageContext.cs
Comment (emgarten) [inline test/TestUtilities/Test.Utility/SimpleTestSetup/SimpleTestPackageContext.cs:67]: 👍

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (emgarten) [inline test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs:49]: In the other PMC test class I've added helper methods for things such as getting and ensuring the console is open. And for verifying there are no errors in the error window. All of this should be done in one place to make these tests easier to manage. So anything that can be shared between these please do it.

With errors for example, if this fails all you will see is Assert.True was false. It doesn't show you what the errors were, and since this might not repro locally it will be hard to figure out what happened.

I also don't think it makes sense in most scenarios to check the error lists, this is probably going to fail due to build related issues since build is called right before. The extra builds here should also be removed if they are being done after all the asserts, they really have no benefit for the tests.

**Review by emgarten — CHANGES_REQUESTED** [review]
minor comments on cleaning up the tests

File: test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs
Comment (PatoBeltran) [reply test/NuGet.Tests.Apex/NuGet.Tests.Apex/NuGetIVsApiTests/NuGetPackageSigningTestCase.cs:49]: @emgarten the PR for the PMC helper methods you mention is still in WIP. I prefer not to block this PR on that... Either change this class as part of that PR or I can make the change once both are merged.

**Comment by PatoBeltran** [comment]
@jainaashish @emgarten @dtivel  I just updated this PR addressing your comments and cleaning up some code in Apex classes

File: test/TestExtensions/NuGet.StaFact/CIOnlyNuGetWpfTheoryAttribute.cs
Comment (rohit21agrawal) [inline test/TestExtensions/NuGet.StaFact/CIOnlyNuGetWpfTheoryAttribute.cs]: you might want to extend from NuGetWpfTheoryAttribute

Comment (PatoBeltran) [reply test/TestExtensions/NuGet.StaFact/CIOnlyNuGetWpfTheoryAttribute.cs]: @rohit21agrawal by doing this I can remove the `AttributeUsage` and `XunitTestCaseDiscoverer`, because they will be inherited, correct?

Comment (rohit21agrawal) [reply test/TestExtensions/NuGet.StaFact/CIOnlyNuGetWpfTheoryAttribute.cs]: depends on whether the attributes themselves have Inherited=true set

Comment (PatoBeltran) [reply test/TestExtensions/NuGet.StaFact/CIOnlyNuGetWpfTheoryAttribute.cs]: I checked and they will, so removing them works perfectly.

**Review by emgarten — APPROVED** [review]

**Review by jainaashish — APPROVED** [review]
Looks good 👍

**Pull request merged**
